crates/hyperdrive-math/src/lp.rs

```diff
@@ -600,6 +621,540 @@ impl State {
+                    active_lp_total_supply,
+                    withdrawal_shares_total_supply,
+                )?;
+
+            // Step 3: If none of the withdrawal shares could be redeemed, then
+            // we're done and we pay out nothing.
+            if withdrawal_shares_redeemed == fixed!(0) {
+                return Err(eyre!(
+                    "Withdrawal shares redeemed from excess idle is zero."
+                ));
+
+            // Step 3: Otherwise if this amount is less than or equal to the amount
+            // of withdrawal shares outstanding, then we're done and we pay out the
+            // full maximum share reserves delta.
+            } else if withdrawal_shares_redeemed <= withdrawal_shares_total_supply {
+                return Ok((withdrawal_shares_redeemed, max_share_reserves_delta));
+
+            // Step 3: Otherwise, all of the withdrawal shares are redeemed, and we
+            // need to calculate the amount of shares the withdrawal pool should
+            // receive.
+            } else {
+                withdrawal_shares_total_supply
+            }
+        };
+
+        // Step 4: Solve for the share proceeds that hold the LP share price
+        // invariant after all of the withdrawal shares are redeemed. If the
+        // calculation returns a share proceeds of zero, we can't pay out
+        // anything.
+        let share_proceeds = self.calculate_distribute_excess_idle_share_proceeds(
+            current_block_timestamp,
+            active_lp_total_supply,
+            withdrawal_shares_total_supply,
+            max_share_reserves_delta,
+            max_iterations,
+        )?;
+        if share_proceeds == fixed!(0) {
+            return Err(eyre!("Share proceeds from excess idle are zero."));
+        }
+
+        // Step 4: If the share proceeds are greater than or equal to the
+        // maximum share reserves delta that was previously calculated, then
+        // we can't distribute excess idle since we ruled out the possibility
+        // of paying out the full maximum share reserves delta in step 3.
+        if share_proceeds >= max_share_reserves_delta {
+            return Err(eyre!(
+                "Share proceeds greater than or equal to the maximum share reserves delta."
+            ));
+        }
+
+        Ok((withdrawal_shares_redeemed, share_proceeds))
+    }
+
+    fn calculate_distribute_excess_idle_share_proceeds(
+        &self,
+        current_block_timestamp: U256,
+        active_lp_total_supply: FixedPoint, // just the number of lp tokens
+        withdrawal_shares_total_supply: FixedPoint,
+        max_share_reserves_delta: FixedPoint,
+        max_iterations: u64,
+    ) -> Result<FixedPoint> {
+        let net_curve_trade =
+            self.calculate_net_curve_trade_from_timestamp(current_block_timestamp)?;
+        // Calculate the LP total supply.
+        let lp_total_supply = active_lp_total_supply + withdrawal_shares_total_supply;
```

Comment:
above we define `lp_total_supply = active_lp_total_supply + withdrawal_shares_total_supply - withdrawal_shares_ready_to_withdraw`. Do we know that `withdrawal_shares_ready_to_withdraw=0` here? If so, can you comment it so we are consistent in our definitions?

Reply:
@jalextowle In solidity we define `lpTotalSupply` as
```solidity
        uint256 lpTotalSupply = _params.activeLpTotalSupply +
            _params.withdrawalSharesTotalSupply;
  ```
  in `calculateDistributeExcessIdleWithdrawalSharesRedeemed()` and `calculateDistributeExcessIdleShareProceeds()`
  
  and as 
  ```solidity
          uint256 lpTotalSupply = _totalSupply[AssetId._LP_ASSET_ID] +
            _totalSupply[AssetId._WITHDRAWAL_SHARE_ASSET_ID] -
            _withdrawPool.readyToWithdraw;
  ```
in `getPoolConfig()` and `_addLiquidity()`

Just double checking that this is intentional.

Reply:
@dpaiton I'll rename these internal ones to something else, so far I've got `lp_total_supply_and_withdrawal_shares_ready_to_withdraw` but I'm willing to workshop it 😂

Reply:
lol `let active_lp_total_supply_plus_withdrawal_shares_total_supply = active_lp_total_supply + withdrawal_shares_total_supply` should be sufficiently descriptive 😆

Reply:
> @jalextowle In solidity we define `lpTotalSupply` as
> 
> ```solidity
>         uint256 lpTotalSupply = _params.activeLpTotalSupply +
>             _params.withdrawalSharesTotalSupply;
> ```
> 
> in `calculateDistributeExcessIdleWithdrawalSharesRedeemed()` and `calculateDistributeExcessIdleShareProceeds()`
> 
> and as
> 
> ```solidity
>         uint256 lpTotalSupply = _totalSupply[AssetId._LP_ASSET_ID] +
>           _totalSupply[AssetId._WITHDRAWAL_SHARE_ASSET_ID] -
>           _withdrawPool.readyToWithdraw;
> ```
> 
> in `getPoolConfig()` and `_addLiquidity()`
> 
> Just double checking that this is intentional.

Yeah this is intentional. We precompute this before `calculate_distribute_excess_idle`:

```
          _totalSupply[AssetId._WITHDRAWAL_SHARE_ASSET_ID] -
          _withdrawPool.readyToWithdraw;
```